src/Infrastructure/Databases/EFRecentContractRepository.cs

```
@@ -0,0 +1,83 @@
+    {
+        _context = context;
+        _logger = logger;
+    }
+
+    private DbSet<User> Users => _context.Users;
+
+    /// <inheritdoc />
+    public void AddRecent(string username, Contract contract)
+    {
+        User user = GetUserByUserName(username);
+
+        RecentlyViewedContract? recentlyViewedContract =
+            user.RecentlyViewContracts.SingleOrDefault(recentContract => recentContract.ContractId == contract.Id);
+
+        // Checks whether to create a new recent contract or just update the viewed time.
+        if (recentlyViewedContract == null)
+        {
+            var recentContract = new RecentlyViewedContract(contract.Id, user.Id);
+            user.RecentlyViewContracts.Add(recentContract);
+        }
+        else
+        {
+            recentlyViewedContract.LastViewed = DateTime.Now;
+        }
+
+        try
+        {
+            _context.SaveChanges();
+        }
+        catch (DataException e)
+        {
+            _logger.LogError("Could not add contract to recently viewed for user to database: {Message}", e.Message);
+        }
+    }
+
+    /// <inheritdoc />
+    public void RemoveRecent(RecentlyViewedContract contractToRemove)
+    {
+        _context.RecentlyViewedContracts.Remove(contractToRemove);
+        _ = _context.SaveChanges();
+        _logger.LogInformation(
+            "Removed contract with id {ContractId} from user with id {UserId}",
+            contractToRemove.ContractId,
+            contractToRemove.UserId);
+    }
+
+    /// <inheritdoc />
+    public IList<RecentlyViewedContract> FetchRecentContracts(string username)
+    {
+        User user = GetUserByUserName(username);
+        return user.RecentlyViewContracts;
+    }
+
+    private User GetUserByUserName(string username)
+    {
+        User user = Users.Include(user => user.RecentlyViewContracts).First(user => user.Name == username);
```

By using `.First()` here you'll throw an exception if there isn't a user that matches. You should instead 1) catch it and throw a user not found exception, or 2) use `.FirstOrDefault()` and if the returned value is null throw a user not found exception.